optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/bi/BiConstraintStream.java

```diff
@@ -198,6 +198,166 @@
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param <C> the type of the third matched fact
+     * @return never null, a stream that matches every pair of A and B where C does not exist for which the
+     * {@link TriJoiner}s are true
+     */
+    default <C> BiConstraintStream<A, B> ifNotExists(Class<C> otherClass, TriJoiner<A, B, C> joiner1,
+            TriJoiner<A, B, C> joiner2) {
+        return ifNotExists(otherClass, new TriJoiner[] {joiner1, joiner2});
+    }
+
+    /**
+     * As defined by {@link #ifNotExists(Class, TriJoiner)}. For performance reasons, indexing joiners must be placed
+     * before filtering joiners.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @param <C> the type of the third matched fact
+     * @return never null, a stream that matches every pair of A and B where C does not exist for which the
+     * {@link TriJoiner}s are true
+     */
+    default <C> BiConstraintStream<A, B> ifNotExists(Class<C> otherClass, TriJoiner<A, B, C> joiner1,
+            TriJoiner<A, B, C> joiner2, TriJoiner<A, B, C> joiner3) {
+        return ifNotExists(otherClass, new TriJoiner[] {joiner1, joiner2, joiner3});
+    }
+
+    /**
+     * As defined by {@link #ifNotExists(Class, TriJoiner)}. For performance reasons, indexing joiners must be placed
+     * before filtering joiners.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @param joiner4 never null
+     * @param <C> the type of the third matched fact
+     * @return never null, a stream that matches every pair of A and B where C does not exist for which the
+     * {@link TriJoiner}s are true
+     */
+    default <C> BiConstraintStream<A, B> ifNotExists(Class<C> otherClass, TriJoiner<A, B, C> joiner1,
+            TriJoiner<A, B, C> joiner2, TriJoiner<A, B, C> joiner3, TriJoiner<A, B, C> joiner4) {
+        return ifNotExists(otherClass, new TriJoiner[] {joiner1, joiner2, joiner3, joiner4});
+    }
+
+    /**
+     * As defined by {@link #ifNotExists(Class, TriJoiner)}. For performance reasons, indexing joiners must be placed
+     * before filtering joiners.
+     * <p>
+     * This method causes <i>Unchecked generics array creation for varargs parameter</i> warnings,
+     * but we can't fix it with a {@link SafeVarargs} annotation because it's an interface method.
+     * Therefore, there are overloaded methods with up to 4 {@link TriJoiner} parameters.
+     * @param otherClass never null
+     * @param joiners never null
+     * @param <C> the type of the third matched fact
+     * @return never null, a stream that matches every pair of A and B where C does not exist for which the
+     * {@link TriJoiner}s are true
+     */
+    <C> BiConstraintStream<A, B> ifNotExists(Class<C> otherClass, TriJoiner<A, B, C>... joiners);
```

Comment:
I just experimented with, this, and if called like `ifNotExists(Visit.class)` - without the joiners - this gives varargs warnings in IDE. Not sure what to do about this method.

Reply:
In the javadoc it says:
     * This method causes <i>Unchecked generics array creation for varargs parameter</i> warnings,
     * but we can't fix it with a {@link SafeVarargs} annotation because it's an interface method.
     * Therefore, there are overloaded methods with up to 4 {@link TriJoiner} parameters.
I've seen that issue on all other varargs methods before - is this the same issue as what you are seeing?

If it is, ignore it. If it's not, let's discuss it.

Reply:
Yeah, it is the same with joins etc. Ok, leaving it as it is.